Implemented multiple folders support

yyoncho opened the pull request:
Fixes https://github.com/emacs-lsp/lsp-mode/issues/345

* Added methods for adding/removing folders
* Tested with JDT server which seems to be the only server supporting the
feature.
* Verified that clients which does not support workspace folders work as
  expected(verfied with lsp-python)

vibhavp commented:
Thanks for the PR! Unfortunately, I've been a little busy with real life, so it might take some time before I get to review this PR.

yyoncho commented:
@vibhavp sure, I will extend the PR description with more information on how it is going to work to help you with the reviewing.

yyoncho commented:
Updated the fix - added some more explanation in the commit description. Fixed corner cases related to deletion of the last element. I have also renamed the function `lsp--get-project-root` to `lsp--suggest-project-root` because the name should not imply that the method returns the proper project root.

yyoncho commented:
@vibhavp did you had a chance to take a look?

yyoncho commented:
Rebased onto master and removed formatting changes to avoid constant merging.

yyoncho commented:
@MaskRay - can you take a look, it looks like @vibhavp is pretty busy these days?

MaskRay commented on lsp-methods.el:
```diff
@@ -855,16 +973,39 @@ directory."
-                                                   extra-init-params)))
-      (setf response (lsp--send-request
-                      (lsp--make-request "initialize" init-params)))
+                                :initializationOptions ,extra-init-params-resolved)
```
indent this line

MaskRay commented on lsp-methods.el:
```diff
@@ -855,16 +973,39 @@ directory."
+       ;; language extension to keep backward compatibility with `lsp-java'
+       workspace-folders (append
+                          (lsp--read-from-file
+                           (concat (file-name-as-directory root) ".folders"))
```
Is it conventional to use `$root/.folders`?

yyoncho replied:
I am open to alternative suggestions.

MaskRay replied:
@yyoncho You might think I can merge PRs in this repo but I can't.. The most I can give you is an approval..

MaskRay commented on lsp-methods.el:
```diff
@@ -819,19 +829,127 @@ directory."
+(defun lsp--workspace-folders-capability-p ()
+  "Return whether current server supports workspace folders."
+  (when-let (workspace (gethash "workspace" (lsp--server-capabilities)))
+    (when-let (capability (gethash "workspaceFolders" workspace))
```
`when-let*` ?

yyoncho replied:
Just checked out when-let* is not available in emacs 25(unless I am mistaken).

MaskRay commented on lsp-methods.el:
```diff
@@ -24,6 +24,12 @@
 (require 'inline)
 (require 'em-glob)
 
+(defcustom lsp-workspace-folders-changed nil
```
How about `on-workspace-folders-change`? `workspace-folders-changed`

yyoncho replied:
Sure, will update the PR.

MaskRay commented:
Is https://github.com/emacs-lsp/lsp-mode/pull/356 superseded by this PR? You may also ask @samrayleung to review

I don't know/write much elisp...

stardiviner commented:
I have an question, does this PR add a feature that user can get lsp-mode works without `lsp--cur-workspace`? For example, I want to get code completion with `company-lsp` in Org Mode source code opened dedicated buffer.

yyoncho commented:
@stardiviner no. Can you share which languages do you want to use in org-mode buffers?

stardiviner commented:
@yyoncho lsp-python and lsp-php and lsp-java.

yyoncho commented:
@stardiviner do you know that should be done on `org-mode` side to do the integration? I am looking at https://stackoverflow.com/questions/14704077/how-to-enable-auto-complete-in-emacs-org-babel is this how it is supposed to work?

stardiviner commented:
Org Mode seems can't implement this. The answers does not work in this case. The first selected answer use `narrow` to create a new indirect buffer that can change `major-mode`. But will mess up original buffer syntax highlighting (font-locking?). The second answer need the Org source block language support session. Also this session support are usually in REPL buffer.

If `lsp-mode` can't support completion without workspace. I think my feature request is not able to implemented.

Sorry for distribution.

yyoncho commented:
@stardiviner I believe that it can work by providing session properties, but I need working example to start with and I can provide a sample which works with LSP mode.

MaskRay commented:
For C/C++ lsp-mode users who want to try out this PR, I have implemented workspace folders in ccls in the weekend.

I'm down for this PR as long as there is still some way to enable automatic root detection which is convenient for browsing random new projects.

yyoncho commented:
@MaskRay can you elaborate?  This CL does not break the current behaviour(ability to browse random new projects), it only adds an option to call add/remove folders for a particular workspace.

inspell commented:
This issue is PITA for `lsp-java`. Could someone from `lsp-mode` take a look?

vibhavp commented:
LGTM, Apologies for the late review. Thanks!
